nucypher/policy/kits.py

```diff
@@ -0,0 +1,247 @@
+        self.signature = signature
+
+    def __eq__(self, other):
+        return (self.ciphertext == other.ciphertext and
+                self.capsule == other.capsule and
+                self.sender_verifying_key == other.sender_verifying_key and
+                self.signature == other.signature)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}({self.capsule})"
+
+    def __bytes__(self):
+        return (bytes(self.capsule) +
+                (b'\x00' if self.signature is None else (b'\x01' + bytes(self.signature))) +
+                (b'\x00' if self.sender_verifying_key is None else (b'\x01' + bytes(self.sender_verifying_key))) +
+                VariableLengthBytestring(self.ciphertext))
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        splitter = BytestringSplitter(
+            capsule_splitter,
+            (bytes, 1))
+
+        capsule, signature_flag, remainder = splitter(data, return_remainder=True)
+
+        if signature_flag == b'\x00':
+            signature = None
+        elif signature_flag == b'\x01':
+            signature, remainder = signature_splitter(remainder, return_remainder=True)
+        else:
+            raise ValueError("Incorrect format for the signature flag")
+
+        splitter = BytestringSplitter(
+            (bytes, 1))
+
+        key_flag, remainder = splitter(remainder, return_remainder=True)
+
+        if key_flag == b'\x00':
+            sender_verifying_key = None
+        elif key_flag == b'\x01':
+            sender_verifying_key, remainder = key_splitter(remainder, return_remainder=True)
+        else:
+            raise ValueError("Incorrect format of the sender's key flag")
+
+        ciphertext, = BytestringSplitter(VariableLengthBytestring)(remainder)
+
+        return cls(capsule, ciphertext, signature=signature, sender_verifying_key=sender_verifying_key)
+
+    def as_policy_kit(self, policy_key, threshold):
+        return PolicyMessageKit.from_message_kit(self, policy_key, threshold)
+
+    def as_retrieval_kit(self):
+        return RetrievalKit(self.capsule, set())
+
+
+class RetrievalKit:
+    """
+    An object encapsulating the information necessary for retrieval of cfrags from Ursulas.
+    Contains the capsule and the checksum addresses of Ursulas from which the requester
+    already received cfrags.
+    """
+
+    def __init__(self, capsule: Capsule, queried_addresses: List[ChecksumAddress]):
+        self.capsule = capsule
+        # Can store cfrags too, if we're worried about Ursulas supplying duplicate ones.
+        self.queried_addresses = queried_addresses
+
+    def __bytes__(self):
+        return (bytes(self.capsule) +
+                b''.join(to_canonical_address(address) for address in self.queried_addresses))
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        capsule, remainder = capsule_splitter(data, return_remainder=True)
+        addresses_as_bytes = checksum_address_splitter.repeat(remainder)
```

Comment:
If it is possible for `queried_addresses` to be an empty list, this line will fail recreating such a `RetrievalKit` from its bytes. Either we prevent empty lists of addresses or handle them.

```python
In [7]: retrieval_kit = RetrievalKit(capsule, [])

In [8]: retrieval_kit_bytes = bytes(retrieval_kit)

In [9]: retrieval_kit_bytes
Out[9]: b'\x02%\x01\xb8f\xd2I\x05\x1a\xb8\xa2\xeb_l\rv\xca\x94\x14\xec\xbaoE\xd7\x0c\xbf\x80\x91_\xb9V\xb5\x93\x03\xd9\x1c\xd7\xa18\t\x11\x0b\xbb\xb3\x06\x0b\x13\x94\xc9\x0b\xe5\xbc\xfc\x1d\xb1X\x8d9#Vi\xe9\xa0\xee\xc2\x85\xfcl\x1f\x9d\x8a\xdd^\xfe\x04`\x9c,*\xb8\xc0\xb1\x8b8\xda[\xe7\x08\x99a\xa3\xf4>\x11\xfc\xbb!\xa1'

In [10]: r = RetrievalKit.from_bytes(retrieval_kit_bytes)
---------------------------------------------------------------------------
BytestringSplittingError                  Traceback (most recent call last)
<ipython-input-10-b61f49bde05f> in <module>
----> 1 r = RetrievalKit.from_bytes(retrieval_kit_bytes)

~/Documents/Github/repos/forks/derek/nucypher/nucypher/policy/kits.py in from_bytes(cls, data)
    157     def from_bytes(cls, data: bytes):
    158         capsule, remainder = capsule_splitter(data, return_remainder=True)
--> 159         addresses_as_bytes = checksum_address_splitter.repeat(remainder)
    160         return cls(capsule, [to_checksum_address(address) for address in addresses_as_bytes])
...
BytestringSplittingError: Not enough bytes to constitute message types bytes:20 - need 20, got 0
```

Reply:
Good point, fixed.